🌐 Added i18n support for comments mails

SebastianSchroeder opened the pull request:
Enables missing i18n support for mails being sent out for new comments, replies to comments or reported comments. Added German translations.

SebastianSchroeder commented on ghost/i18n/locales/vi/ghost.json:
```diff
@@ -1,5 +1,7 @@
 {
     "{date}": "{date}",
+    "A comment has been reported on {postTitle}": "",
+    "A comment has been reported on your post": "",
```
Need to be added in a different PR

SebastianSchroeder commented on ghost/core/core/server/services/comments/email-templates/new-comment-reply.hbs:
```diff
@@ -166,20 +166,20 @@
 
                     <!-- START FOOTER -->
                     <tr>
                       <td style="font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol'; font-size: 14px; vertical-align: top; padding-top: 80px;">
```
That would be a good way to do it. The html around it with the style information would need to be included in the translation key. Are you suggestion to do that?

cathysarisky replied:
@SebastianSchroeder , CodeRabbit's suggestion should work here, and avoids adding a new dependency (@doist/Interpolate) like how the problem was handled in Portal.

SebastianSchroeder commented on ghost/core/core/server/services/comments/email-templates/new-comment-reply.hbs:
```diff
@@ -120,8 +120,8 @@
                       <td style="font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol'; font-size: 14px; vertical-align: top;">
-                        <p style="font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol'; font-size: 20px; color: #15212A; font-weight: bold; line-height: 25px; margin: 0; margin-bottom: 15px;">Hey there,</p>
-                        <p style="font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol'; font-size: 16px; color: #3A464C; font-weight: normal; margin: 0; line-height: 25px; margin-bottom: 32px;">Someone just replied to your comment on <a href="{{postUrl}}" target="_blank" style="font-weight: bold; ...
+                        <p style="font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol'; font-size: 20px; color: #15212A; font-weight: bold; line-height: 25px; margin: 0; margin-bottom: 15px;">{{t 'Hey there,'}}</p>
+                        <p style="font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol'; font-size: 16px; color: #3A464C; font-weight: normal; margin: 0; line-height: 25px; margin-bottom: 32px;">{{t 'Someone just replied to your comment on'}} <a href="{{postUrl}}" target="_blank" style="font-weigh ...
```
You omitted the style information. I would need to include them as well, right?

cathysarisky reviewed (commented):
Hey @SebastianSchroeder , thank you for identifying another spot for translations! I appreciate your diligence!

I've left you a couple comments on this PR in the limited time I have available this morning.  It needs a fuller review still, but here are a couple of general thoughts that might allow you to make some more progress:

- The current overall i18n goal is to make the members interface 100% translatable.  The admin interface is not a goal for translation right now, as far as I know.  A significant part of this work is actually targeting emails that go to staff members.  A more focused PR that only addresses emails that go to members (that's the comment replies) would probably get approved more rapidly.  [I'm going to check in with the team on this and will report back.]

- It's going to be important to wrap text for translation into logical chunks.  Some of the wrapping breaks text up into smaller pieces that may not be in the right order in all languages.  Generally, wrapping the full text at sentence size gives translators the most flexibility.  Coderabbit gave you a good suggestion for how to handle the formatting.  It looks a little ugly in the code, but it completely hides the html from the translators.

- You're going to need more tests. All the current tests need to pass (they aren't currently), and then you need tests that show that the translation is occurring when the language is not English.  In your test set, be sure to include a case where you pass in a non-language code (like 'xx' and '') and confirm that English emails are sent -- I learned during newsletter translation that some sites have all sorts of nonsense in their locale field, so your code needs to deal with that.

SebastianSchroeder committed:
🌐 Added i18n support for comments mails

Enables missing i18n support for mails being sent out for new comments, replies to comments or reported comments. Added German translations.

SebastianSchroeder commented:
Hello @cathysarisky,

thank you for the helpful feedback.

I removed the translations for the author emails and only kept the reply to comment email translations.
I applied the suggestions from Coderabbit which made the translation texts a lot cleaner.

As a last step I will fix the failing tests and add additional tests for the new translations.

SebastianSchroeder committed:
Fix typo in CommentsServiceEmails.js

SebastianSchroeder commented:
@cathysarisky I added tests to assert the i18n functionality of the email rendering. Could you have a look at those and tell me if they are in line with what you would expect? I extracted the rendering into a new class as that made testing a lot easier. If the tests are approved I would add more tests for the i18n of the subject of the mail.

What I did not test is illegal locale values. Do I need to check for those every time I use the i18n module? Should that not be a global check rather than every usage of the i18n having to assert that?
I looked at ghost/core/core/server/services/email-service/EmailServiceWrapper.js:61 for example. The lines seem to duplicate what is done in ghost/core/core/server/services/i18n.js. The is not a check against illegal values but maybe an example if functionality is duplicated when a global module is being used.

SebastianSchroeder commented:
> The reason I suggest checking with invalid values is that the locale value is set by the user in the admin dashboard. So you can't rely on it being among the list of locales, or not being blank, or much of anything else. If any function you're calling doesn't gracefully fall back to English when the locale is invalid, that's going to be important to catch.

I understand your reasoning. What I was wondering is if the fallback to English would not be better implemented in the i18n module instead of in every module using the i18n module?

cathysarisky commented:
> > (I got into trouble last year with email newsletters -- I had a spot where a date translation failed in a way that stopped newsletters from being sent entirely. That was not a good day. So now I always test.)
> 
> @cathysarisky are you able to provide a reference to that? My understanding is that `i18n` will [default to `'en'`](https://github.com/TryGhost/Ghost/blob/main/ghost/i18n/lib/i18n.js#L112) when translations for the requested language are not available but maybe I'm missing something. Having a concrete example of the type of problem you're referring to would help verify the code and that tests are checking against the right thing.

@kevinansfield , thanks for reviewing!  The case I was referring to was the one patched here: https://github.com/TryGhost/Ghost/pull/21573/files

cathysarisky reviewed (approved):
As Kevin approves, I am marking my review as approved also!

SebastianSchroeder commented:
Thank you for your comments. I'll remove the test for the invalid locale.

Testing the `notifyPostAuthors()` would have been nice but as there are no tests for that yet it would have been quite a lot of work to add them causing more refactoring.

cathysarisky commented on ghost/core/test/unit/server/services/comments/comments-service-emails-renderer.test.js:
```diff
@@ -0,0 +1,47 @@
+            should(result.text).containEql('Hey there,');
+            should(result.text).containEql('Someone just replied to your comment on Test Post.');
+        });
+
```
```suggestion

        it('should correctly handle apostrophes in post titles and site names', async function () {
            // arrange
            const i18n = i18nLib('en', 'ghost');
            const renderer = new CommentsServiceEmailRenderer({t: i18n.t});

            const templateData = {
                postTitle: 'Test post\'s the best post',
                siteUrl: 'https://ghost.org',
                siteTitle: 'Cathy\'s blog'
                
            };

            // act
            const result = await renderer.renderEmailTemplate('new-comment-reply', templateData);

            // assert
            should(result.html).containEql('Hey there,</p>');
            should(result.html).containEql('This message was sent from <a class="small" href="https://ghost.org" style="text-decoration: underline; color: #738A94; font-size: 11px;">Cathy\'s blog</a>');
            should(result.text).containEql('Someone just replied to your comment on Test post\'s the best post.');
        });
        
```

cathysarisky reviewed (commented):
one test suggestion.

cathysarisky committed:
Add a test for double-encoding

cathysarisky commented:
Thanks @SebastianSchroeder !  I'm pushing one additional test and then will get this merged today!

SebastianSchroeder commented:
Your test actually revealed a bug. I'll fix it now.

SebastianSchroeder committed:
Fix malformed template variables and extend test

SebastianSchroeder commented:
The Handlebars `registerHelper` sets the escapeValue to false for all usages of t in the handlebars templates. It basically creates a new t function. 
If the helper is used in other places the escapeValue has to be set again. Does that make sense or am I missing something?

cathysarisky commented:
Oh, you're totally right.  Sorry, I got briefly muddled between the handlebars t and the i18n function t.  Possibly because I haven't yet had enough t(ea) this morning. 🤣 Merging now!
